src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs

```diff
@@ -0,0 +1,546 @@
+        ///         <item>For entries where the language encoding flag (EFS) in the general purpose bit flag of the local file header <em>is</em> set,
+        ///         use UTF-8 (<c>Encoding.UTF8</c>) in order to decode the entry name.</item>
+        ///     </list>
+        ///     <para>Note that Unicode encodings other than UTF-8 may not be currently used for the <c>entryNameEncoding</c>,
+        ///     otherwise an <see cref="ArgumentException"/> is thrown.</para>
+        /// </param>
+        public static void ExtractToDirectory(String sourceArchiveFileName, String destinationDirectoryName, Encoding entryNameEncoding)
+        {
+            if (sourceArchiveFileName == null)
+                throw new ArgumentNullException("sourceArchiveFileName");
+
+            using (ZipArchive archive = Open(sourceArchiveFileName, ZipArchiveMode.Read, entryNameEncoding))
+            {
+                archive.ExtractToDirectory(destinationDirectoryName);
+            }
+        }
+
+
+        private static void DoCreateFromDirectory(String sourceDirectoryName, String destinationArchiveFileName,
+                                                  CompressionLevel? compressionLevel, Boolean includeBaseDirectory,
+                                                  Encoding entryNameEncoding)
+        {
+            // Rely on Path.GetFullPath for validation of sourceDirectoryName and destinationArchive
+
+            // Checking of compressionLevel is passed down to DeflateStream and the IDeflater implementation
+            // as it is a pluggable component that completely encapsulates the meaning of compressionLevel.
+
+            sourceDirectoryName = Path.GetFullPath(sourceDirectoryName);
+            destinationArchiveFileName = Path.GetFullPath(destinationArchiveFileName);
+
+            using (ZipArchive archive = Open(destinationArchiveFileName, ZipArchiveMode.Create, entryNameEncoding))
+            {
+                bool directoryIsEmpty = true;
+
+                //add files and directories
+                DirectoryInfo di = new DirectoryInfo(sourceDirectoryName);
+
+                string basePath = di.FullName;
+
+                if (includeBaseDirectory && di.Parent != null)
+                    basePath = di.Parent.FullName;
+
+                foreach (FileSystemInfo file in di.EnumerateFileSystemInfos("*", SearchOption.AllDirectories))
+                {
+                    directoryIsEmpty = false;
+
+                    Int32 entryNameLength = file.FullName.Length - basePath.Length;
+                    Contract.Assert(entryNameLength > 0);
+
+                    String entryName = file.FullName.Substring(basePath.Length, entryNameLength);
+
+                    // Remove any leading slashes from the entry name:
+                    entryName = entryName.TrimStart(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
```

Comment:
It looks as though entryName is supposed to be the name of the entry we create in the archive

archive.CreateEntry(entryName + Path.DirectorySeparatorChar);

In that case it makes sense these would be relative to the archive.